tests/misc-utils.c

```diff
@@ -0,0 +1,295 @@
+{
+  GQueue queue = G_QUEUE_INIT;
+  GList *dir_stack = NULL;
+  g_autoptr(GError) local_error = NULL;
+  g_autoptr(GFileInfo) top_info;
+
+  top_info = g_file_query_info (topdir,
+                                G_FILE_ATTRIBUTE_STANDARD_TYPE,
+                                G_FILE_QUERY_INFO_NOFOLLOW_SYMLINKS,
+                                NULL,
+                                &local_error);
+  if (top_info == NULL)
+    {
+      if (g_error_matches (local_error, G_IO_ERROR, G_IO_ERROR_NOT_FOUND))
+        return TRUE;
+      g_propagate_error (error, g_steal_pointer (&local_error));
+      return FALSE;
+    }
+  if (g_file_info_get_file_type (top_info) != G_FILE_TYPE_DIRECTORY)
+    return g_file_delete (topdir, NULL, error);
+
+  g_queue_push_head (&queue, g_object_ref (topdir));
+  dir_stack = g_list_prepend (dir_stack, g_object_ref (topdir));
+  while (!g_queue_is_empty (&queue))
+    {
+      g_autoptr(GFile) dir = G_FILE (g_queue_pop_tail (&queue));
+      g_autoptr(GFileEnumerator) enumerator = g_file_enumerate_children (dir,
+                                                                         G_FILE_ATTRIBUTE_STANDARD_TYPE,
+                                                                         G_FILE_QUERY_INFO_NOFOLLOW_SYMLINKS,
+                                                                         NULL,
+                                                                         error);
+
+      if (enumerator == NULL)
+        return FALSE;
+
+      for (;;)
+        {
+          GFileInfo *info;
+          GFile* child;
+
+          if (!g_file_enumerator_iterate (enumerator,
+                                          &info,
+                                          &child,
+                                          NULL,
+                                          error))
+            return FALSE;
+
+          if (info == NULL || child == NULL)
+            break;
+
+          if (g_file_info_get_file_type (info) == G_FILE_TYPE_DIRECTORY)
+            {
+              g_queue_push_head (&queue, g_object_ref (child));
+              dir_stack = g_list_prepend (dir_stack, g_object_ref (child));
+            }
+          else if (!g_file_delete (child, NULL, error))
+            return FALSE;
+        }
+    }
+
+  while (dir_stack != NULL)
+    {
+      GList *first = dir_stack;
+      g_autoptr(GFile) dir = G_FILE (first->data);
+
+      dir_stack = g_list_remove_link (dir_stack, first);
+      g_list_free_1 (first);
+      if (!g_file_delete (dir, NULL, error))
+        return FALSE;
+    }
+
+  return TRUE;
+}
+
+gchar **
+generate_strv (const gchar *str, ...)
+{
+  g_autoptr(GPtrArray) vec = string_array_new ();
+  va_list args;
+
+  va_start (args, str);
+  while (str != NULL)
+    {
+      g_ptr_array_add (vec, g_strdup (str));
+      str = va_arg (args, const gchar *);
+    }
+  va_end (args);
+  g_ptr_array_add (vec, NULL);
+
+  return (gchar**)g_ptr_array_free (g_steal_pointer (&vec), FALSE);
```

Comment:
I'm trying to recall why I did it that way. It is probably because of having this `SSDEF` and `SS`  hack. Basically `char ** a = {"a", "b", "c"}` might be fine (or rather could be fine if it ended with `NULL`, like `{"a", "b", "c", NULL}`), but the problem was that you couldn't call `g_strfreev` (or use `g_auto(GStrv)`) on something like `{g_strdup_printf("%d", 42), g_strconcat ("b", "c", NULL), NULL }`. Instead you would need to hold separate variables for the strings, which gets boring pretty quickly. Also, `generate_strv` works on something like `("a", SS (g_strdup_printf (…), NULL)`.